models/record.go

```diff
@@ -0,0 +1,235 @@
+package models
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strings"
+
+	"github.com/miekg/dns"
+	"github.com/miekg/dns/dnsutil"
+)
+
+// RecordConfig stores a DNS record.
+// Valid types:
+//   Official:
+//     A
+//     AAAA
+//     ANAME
+//     CAA
+//     CNAME
+//     MX
+//     NS
+//     PTR
+//     SRV
+//     TLSA
+//     TXT
+//   Pseudo-Types:
+//     ALIAs
+//     CF_REDIRECT
+//     CF_TEMP_REDIRECT
+//     FRAME
+//     IMPORT_TRANSFORM
+//     NAMESERVER
+//     NO_PURGE
+//     PAGE_RULE
+//     PURGE
+//     URL
+//     URL301
+type RecordConfig struct {
+	Type             string            `json:"type"`
+	Name             string            `json:"name"`   // The short name. See above.
+	Target           string            `json:"target"` // If a name, must end with "."
+	TTL              uint32            `json:"ttl,omitempty"`
+	Metadata         map[string]string `json:"meta,omitempty"`
+	NameFQDN         string            `json:"-"` // Must end with ".$origin". See above.
+	MxPreference     uint16            `json:"mxpreference,omitempty"`
+	SrvPriority      uint16            `json:"srvpriority,omitempty"`
+	SrvWeight        uint16            `json:"srvweight,omitempty"`
+	SrvPort          uint16            `json:"srvport,omitempty"`
+	CaaTag           string            `json:"caatag,omitempty"`
+	CaaFlag          uint8             `json:"caaflag,omitempty"`
+	TlsaUsage        uint8             `json:"tlsausage,omitempty"`
+	TlsaSelector     uint8             `json:"tlsaselector,omitempty"`
+	TlsaMatchingType uint8             `json:"tlsamatchingtype,omitempty"`
+	TxtStrings       []string          `json:"txtstrings,omitempty"` // TxtStrings stores all strings (including the first). Target stores only the first one.
+	R53Alias         map[string]string `json:"r53_alias,omitempty"`
+
+	CombinedTarget bool `json:"-"`
+
+	Original interface{} `json:"-"` // Store pointer to provider-specific record object. Used in diffing.
+}
+
+// Providers are responsible for validating or normalizing the data
+// that goes into a RecordConfig.
+
+// Name:
+//    This is the shortname i.e. the NameFQDN without the origin suffix.
+//    It should never have a trailing "."
+//    It should never be null. The apex (naked domain) is stored as "@".
+//    If the origin is "foo.com." and Name is "foo.com", this literally means
+//        the intended FQDN is "foo.com.foo.com." (which may look odd)
+// NameFQDN:
+//    This is the FQDN version of Name.
+//    It should never have a trailiing ".".
+// NOTE: Eventually we will unexport Name/NameFQDN. Please start using
+//     the setters (SetLabel/SetLabelFQDN) and getters (Label/LabelFQDN).
+
+// Target:
+//   This is
+
+// SetLabel sets the .Name/.NameFQDN fields given a short name and origin.
+// origin may not have a trailing dot.
```

Comment:
I think  it would be better to say `origin MUST not have a trailing dot.`

Reply:
Done